src/Sentry.Profiling/SampleProfilerSession.cs

```diff
@@ -0,0 +1,46 @@
+using System.Diagnostics.Tracing;
+using Microsoft.Diagnostics.NETCore.Client;
+using Microsoft.Diagnostics.Tracing.EventPipe;
+using Microsoft.Diagnostics.Tracing.Parsers;
+
+namespace Sentry.Profiling;
+
+internal class SampleProfilerSession
+{
+    private DiagnosticsClient _client;
+    private EventPipeSession _session;
+    private MemoryStream _stream = new MemoryStream();
+    private Task _copyTask;
+    private CancellationToken _cancellationToken;
+    private CancellationTokenRegistration _cancellationTokenStopRegistration;
+
+    public SampleProfilerSession(CancellationToken cancellationToken)
+    {
+        _cancellationToken = cancellationToken;
+        _client = new DiagnosticsClient(Process.GetCurrentProcess().Id);
+
+        var providers = new[]
+        {
+            // Note: all events we need issued by "DotNETRuntime" provider are at "EventLevel.Informational"
+            // see https://learn.microsoft.com/en-us/dotnet/fundamentals/diagnostics/runtime-events
+            new EventPipeProvider("Microsoft-Windows-DotNETRuntime", EventLevel.Informational, (long)ClrTraceEventParser.Keywords.Default),
+            new EventPipeProvider("Microsoft-DotNETCore-SampleProfiler", EventLevel.Informational),
+            new EventPipeProvider("System.Threading.Tasks.TplEventSource", EventLevel.Informational, (long)TplEtwProviderTraceEventParser.Keywords.Default)
+        };
+
+        // The size of the runtime's buffer for collecting events in MB, same as the current default in StartEventPipeSession().
+        var circularBufferMB = 256;
+        _session = _client.StartEventPipeSession(providers, true, circularBufferMB);
+        _copyTask = _session.EventStream.CopyToAsync(_stream, _cancellationToken);
+        _cancellationTokenStopRegistration = _cancellationToken.Register(() => _session.Stop(), false);
```

Comment:
> If I understand correctly, this is going to start a profiling session and start copying data to the stream immediately.
> 
> Is there risk that `_copyTask` will complete prematurely? Or is it guaranteed to keep streaming until it is awaited?

I'm not aware either way - it's the standard `Stream.CopyToAsync()`. What would we need to change either way?

> Also, I tend to prefer not starting actual work from within a constructor. Constructors should generally create objects without having side effects (starting tasks, performing file or network i/o, etc.). We have a few violations of this already (`BackgroundWorker`, for example) - but I don't want to continue this anti-pattern. The better approach is to use a static factory method. Thanks.

OK, changed added `static SampleProfilerSession.StartNew()` and `SamplingTransactionProfiler.Start()`

Reply:
The factory methods looks great.  Thanks!

On `CopyToAsync` - that will copy until reaching the end of the stream.  But I'm not sure if the end of the stream means when it read zero bytes from the source stream, or if it will keep at it until the source stream is closed.  I'll check and get back to you.

Reply:
Yeah, I think the copy is problematic.  There are two problems:

- As suspected, copying will stop at the end of the stream, but that is determined by zero bytes remaining to be read from the source stream - not by the stream being closed.
- Copying starts based on the position of the stream at the time the copy method is invoked.  Writing bytes to the source stream will move the position to the end.

Consider this small test program:

```csharp
using var source = new MemoryStream();
using var destination = new MemoryStream();

// this byte will copy to the destination stream
source.WriteByte(1);

// ... but only if the position is reset before copying
source.Position = 0;

var copyTask = source.CopyToAsync(destination);
Console.WriteLine(copyTask.Status);

// this will not copy, because the task has already ran to completion
source.WriteByte(2);

await copyTask;

Dump(source);
Dump(destination);

void Dump(MemoryStream stream)
{
    var bytes = stream.GetBuffer()[..(int) stream.Length];
    Console.WriteLine($"[{string.Join(',', bytes)}]");
}
```

Reply:
While this applies to `MemoryStream`, it doesn't imply the same is true for the type of stream that is `_session.EventStream`. The `CopyToAsync()` method there seems to be implemented differently. Also see the same pattern here: https://github.com/dotnet/diagnostics/blob/main/documentation/diagnostics-client-library-instructions.md#4-trigger-a-cpu-trace-for-given-number-of-seconds

Reply:
Ok.  That looks to be a `NamedPipeClientStream`, which isn't seekable - so that addresses the position issue.  I think the rest of it is working as a side effect of the `EventPipeSession` starting first.  It must start writing to the stream immediately and continuously.  So when copying will continue until after the session is stopped, and all the data from that session is read.

This will work fine I think.  Thanks for walking through it with me. 👍